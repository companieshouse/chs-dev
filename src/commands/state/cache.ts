import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
