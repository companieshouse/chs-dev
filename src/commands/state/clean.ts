import { Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Clean extends Command {

    static description = "Wipes the state of chs-dev and revert to a default state";

    static flags = {
        purge: Flags.boolean({
            char: "p",
            description: "Wipe state, volumes and images",
            default: false
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { flags: { purge } } = await this.parse(Clean);

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`This will reset your chs-dev state and delete all unused docker volumes${purge ? " including images." : "."} Do you want to proceed?`);
        };

        if (await handlePrompt()) {
            const containerStatus = this.dockerCompose.getServiceStatuses();
            if (containerStatus !== undefined) {
                this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
            }
            this.stateManager.cleanState();

            await this.config.runHook("generate-runnable-docker-compose", {});

            this.dockerCompose.prune("volume");
            if (purge) {
                this.dockerCompose.prune("image");
            }

        }
    }

}
