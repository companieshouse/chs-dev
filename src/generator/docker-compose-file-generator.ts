import { existsSync, mkdirSync, readFileSync, rmSync } from "fs";
import { EOL } from "os";
import { join } from "path";
import yaml from "yaml";
import { deduplicate } from "../helpers/array-reducers.js";
import { getInitialDockerComposeFile } from "../helpers/docker-compose-file.js";
import { getAllFilesInDirectory } from "../helpers/file-utils.js";
import { DockerComposeSpec } from "../model/DockerComposeSpec.js";
import { Service } from "../model/Service.js";
import { getBuilder } from "../state/builders.js";
import DevelopmentDockerComposeSpecFactory from "./development/development-docker-compose-factory.js";
import ExclusionDockerComposeSpecFactory from "./exclusion/exclusion-docker-compose-factory.js";
import { AbstractFileGenerator } from "./file-generator.js";
import { RunnableServicesObject, ServiceWithLiveUpdate } from "./interface/index.js";

export class DockerComposeFileGenerator extends AbstractFileGenerator {

    constructor (path: string) {
        super(path, "docker-compose.yaml");
    }

    /**
     * Generates the main Docker Compose file in the root directory based on the service configurations.
     * Add ingress-proxy service and its dependencies.
     * @param services - List of all services with live updates.
     * @param  {runnableObjects} - a List of runnable services and infrastrutural services generated from an exclusion execution.
     * @param hasExcludedServices - Indicates if there are excluded services.
     */
    generateDockerComposeFile (services: ServiceWithLiveUpdate[], runnableObjects = {} as RunnableServicesObject, hasExcludedServices = false) {
        const { runnableServices: runnableServicesList, infrastructureSources } = runnableObjects;
        let dockerCompose = getInitialDockerComposeFile(this.path);
        const runnableServices = runnableServicesList ?? services;

        dockerCompose = this.handleIncludePropertiesInMainDockerCompose(dockerCompose, { runnableServices, infrastructureSources }, hasExcludedServices);

        // Setup the ingress-proxy's dependencies (i.e. every service with an ingress route)
        dockerCompose.services["ingress-proxy"].depends_on = this.listIngressDependencies(runnableServices);

        this.writeFile(
            yaml.stringify(dockerCompose).split(EOL),
            EOL
        );
    }

    /**
     * Generates a development-specific Docker Compose file for a service.
     * @param service - The service for which the file is generated.
     * @param builderVersion - Optional builder version.
     * @param excludedServices - List of excluded service names.
     */
    // Create the development compose file
    generateDevelopmentServiceDockerComposeFile (service: Service, builderVersion: string | undefined, excludedServices: string[] = []) {
        const developmentServicePath = join(this.path, "local", service.name);

        if (!existsSync(developmentServicePath)) {
            mkdirSync(developmentServicePath);
        }

        const developmentComposeFile = join("local", service.name, "docker-compose.yaml");

        const builderDockerComposeSpec = getBuilder(this.path, service.builder, builderVersion);
        const dockerComposeSpecFactory = new DevelopmentDockerComposeSpecFactory(
            builderDockerComposeSpec, this.path
        );

        let dockerComposeConfig = yaml.parse(readFileSync(service.source).toString("utf-8"));

        if (excludedServices.length) {
            const exclusionFactory = new ExclusionDockerComposeSpecFactory();
            dockerComposeConfig = exclusionFactory.removeExcludedServiceFromDependOn(dockerComposeConfig, excludedServices);
        }

        const developmentDockerComposeSpec = dockerComposeSpecFactory.create(
            dockerComposeConfig, service
        );

        // Create the development compose file
        this.writeFile(
            yaml.stringify(developmentDockerComposeSpec).split(EOL),
            EOL,
            developmentComposeFile
        );
    }

    /**
     * Generates exclusion filtered Docker Compose files in the `exclusion-filtered-compose-files` directory and updates the main docker-compose.yaml file `includes` property as a list of the generated exclusion filtered docker-compose files.
     * create exlusion filtered docker-compose for non-development mode services.
     * update development-mode services docker-compose file by removing excluded services from its depend_on property.
     * update main docker-compose file `include` and `depends_on` properties appropriately.
     * @param services - List of all services with live updates.
     * @param excluded - List of excluded service names.
     */
    generateExclusionDockerComposeFiles (services: ServiceWithLiveUpdate[], excluded: string[] | undefined) {
        const exclusionDirPath = join(this.path, "exclusion-runnable-services");

        if (!excluded || !excluded?.length) {
            this.generateDockerComposeFile(services, {} as RunnableServicesObject);
            rmSync(exclusionDirPath, { recursive: true, force: true });
            return;
        }

        if (existsSync(exclusionDirPath)) {
            rmSync(exclusionDirPath, { recursive: true, force: true });
        }
        mkdirSync(exclusionDirPath);

        const exclusionFactory = new ExclusionDockerComposeSpecFactory();

        const { runnableServices, infrastructureSources } = exclusionFactory.handleExcludedServices(services, excluded);

        runnableServices.forEach(service => {
            if (service.liveUpdate) {
                // Create services in dev mode
                this.generateDevelopmentServiceDockerComposeFile(service, undefined, excluded);
            } else {
                // Create services not in dev mode
                const { runnableServicesComposeFile, updatedYamlContent } =
            exclusionFactory.generateDockerComposeFileForExclusionRunnableServices(service, excluded);

                this.writeFile(
                    updatedYamlContent.split(EOL),
                    EOL,
                    runnableServicesComposeFile
                );
            }
        });

        // update the dockercompose on root
        this.generateDockerComposeFile(services, { runnableServices, infrastructureSources }, true);

    }

    /**
     * Lists the ingress dependencies for a given list of services, identifying services that have an ingressRoute defined.
     * It also determines the condition (either 'service_started' or 'service_healthy') based on the presence of a healthcheck.
     *
     * @param services - The list of services to examine.
     * @returns An object where the keys are the service names and the values are their ingress conditions and restart settings.
     */
    private listIngressDependencies (services: ServiceWithLiveUpdate[]): Record<string, {condition: string, restart: boolean}> {
        return services
            .reduce((dependencySpec: Record<string, {condition: string, restart: boolean}>, service) => {
                // Selects services which have an ingressRoute defined
                if ((typeof service.metadata.ingressRoute !== "undefined" && service.metadata.ingressRoute !== null)) {
                    // Condition is selected based upon the presence of healthcheck - when present it will wait for
                    // the service to be healthy as opposed to just started
                    dependencySpec[service.name] = {
                        condition: (typeof service.metadata.healthcheck === "undefined" || service.metadata.healthcheck === null) ? "service_started" : "service_healthy",
                        restart: true
                    };
                }

                return dependencySpec;
            }, {});
    }

    /**
     * Modifies the main Docker Compose configuration by modifing the include property based on whether excluded services exist.
     * If there are excluded services, it includes Compose files for both local and exclusion services.
     * If no exclusions, it only includes the relevant source paths.
     *
     * @param dockerCompose - The original Docker Compose configuration.
     * @param runnableServices - The list of runnable services.
     * @param infrastructuralServicesSources - The sources for infrastructural services.
     * @param hasExcludedServices - A flag indicating whether there are excluded services.
     * @returns The modified Docker Compose configuration with the appropriate services included.
     */
    private handleIncludePropertiesInMainDockerCompose (dockerCompose:DockerComposeSpec, { runnableServices, infrastructureSources: infrastructuralServicesSources }:RunnableServicesObject, hasExcludedServices): DockerComposeSpec {
        if (hasExcludedServices) {
            const localServicesCompose = runnableServices
                .filter(service => service.liveUpdate)
                .map(service => join(this.path, "local", service.name, "docker-compose.yaml"));

            const genratedServicesCompose = getAllFilesInDirectory(join(this.path, "exclusion-runnable-services"));

            dockerCompose.include = [...localServicesCompose, ...genratedServicesCompose, ...infrastructuralServicesSources];
        } else {
            dockerCompose.include = runnableServices.map(
                service => service.liveUpdate
                    ? join(this.path, "local", service.name, "docker-compose.yaml")
                    : service.source
            ).reduce(deduplicate, []);
        }
        return dockerCompose;
    }
}
