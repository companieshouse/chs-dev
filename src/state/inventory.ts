import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { basename, join } from "path";

import yaml from "yaml";
import glob from "glob";

import { createHash } from "crypto";
import { Service } from "../model/Service.js";
import { Module } from "../model/Module.js";
import { readServices } from "./service-reader.js";
import { DependencyNameResolver } from "./dependency-name-resolver.js";

interface InventoryCache {
  hash: string;
  services: Service[];
  modules: Module[]
}

export class Inventory {
    readonly inventoryCacheFile: string;
    private __serviceFiles: string[] | undefined;

    constructor (private path: string, cacheDir: string) {
        this.path = path;
        this.inventoryCacheFile = join(cacheDir, `${basename(path)}.inventory.yaml`);

        if (!existsSync(cacheDir)) {
            mkdirSync(cacheDir, {
                recursive: true
            });
        }
    }

    get modules (): Module[] {
        return this.getFromCache(inventoryCache => inventoryCache.modules);
    }

    get services (): Service[] {
        return this.getFromCache(inventoryCache => inventoryCache.services);
    }

    private getFromCache<T> (cacheSupplier: (inventoryCache: InventoryCache) => T): T {
        const inventoryCache = this.inventoryCache;

        if (typeof inventoryCache !== "undefined") {
            if (inventoryCache.hash === this.hashServiceFiles()) {
                return cacheSupplier(inventoryCache);
            }
        }

        this.updateCache();

        return this.getFromCache(cacheSupplier);
    }

    private updateCache () {
        const inventory = {
            hash: this.hashServiceFiles(),
            services: this.loadServices(),
            modules: this.loadModules()
        };

        writeFileSync(this.inventoryCacheFile, yaml.stringify(inventory));
    }

    private loadModules (): Module[] {
        return readdirSync(join(this.path, "services", "modules"), { withFileTypes: true })
            .filter(item => item.isDirectory())
            .map(item => {
                return {
                    name: item.name
                };
            });
    }

    private loadServices (): Service[] {
        const partialServices: Partial<Service>[] = this.serviceFiles.flatMap(readServices);

        const dependencyNameResolver = new DependencyNameResolver(partialServices);

        return partialServices.map(service => ({
            ...service,
            dependsOn: dependencyNameResolver.fullDependencyListIncludingTransitive(service.dependsOn as string[])
        } as Service));
    }

    private hashServiceFiles () {
        const sha256Hash = createHash("sha256");

        this.serviceFiles
            .map(serviceFile => readFileSync(serviceFile))
            .forEach((serviceFile:any) => {
                sha256Hash.update(serviceFile);
            });

        return sha256Hash.digest("hex");
    }

    private get inventoryCache (): InventoryCache | undefined {
        if (existsSync(this.inventoryCacheFile)) {
            return yaml.parse(readFileSync(this.inventoryCacheFile).toString("utf-8"));
        }

        return undefined;
    }

    private get serviceFiles (): string[] {
        if (typeof this.__serviceFiles === "undefined") {
            this.__serviceFiles = glob.sync(`${this.path}/services/**/*.docker-compose.yaml`);
        }
        return this.__serviceFiles;
    }
}
