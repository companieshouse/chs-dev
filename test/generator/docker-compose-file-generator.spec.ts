import { afterAll, beforeAll, expect } from '@jest/globals'

// @ts-expect-error does exist despite ts warning
import { copyFileSync, existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from 'fs';
import { DockerComposeFileGenerator } from '../../src/generator/docker-compose-file-generator'
import { Service } from '../../src/state/inventory';
import { join } from 'path';
import { parse, stringify } from 'yaml';

describe("DockerComposeFileGenerator", () => {
    let tempDir: string;
    let dockerComposeFileGenerator: DockerComposeFileGenerator;

    beforeAll(() => {
        tempDir = mkdtempSync("docker-compose-file-gen")

        mkdirSync(join(tempDir, "local"))
    })

    afterAll(() => {
        rmSync(tempDir, { recursive: true, force: true });
    })

    beforeEach(() => {
        dockerComposeFileGenerator = new DockerComposeFileGenerator(tempDir)
        mkdirSync(join(tempDir, "services/infrastructure/"), { recursive: true });

        copyFileSync(
            join(process.cwd(), "test/data/docker-compose-file-generator/docker-compose.yaml"),
            join(tempDir, "services/infrastructure/docker-compose.yaml")
        );
    })

    describe("generateDockerComposeFile", () => {

        const services: (Service & {
            liveUpdate: boolean
        })[] = [
                {
                    name: 'service-one',
                    module: 'module-one',
                    source: './services/modules/module-one/service-one.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'java',
                    metadata: {
                        ingressRoute: "Path('One')"
                    },
                    repository: null,
                    liveUpdate: false
                },
                {
                    name: 'service-two',
                    module: 'module-one',
                    source: './services/modules/module-one/service-two.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'node',
                    metadata: {
                        ingressRoute: 'Path("two")'
                    },
                    repository: null,
                    liveUpdate: false
                },
                {
                    name: 'service-three',
                    module: 'module-two',
                    source: './services/modules/module-two/service-three.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'repository',
                    metadata: {},
                    repository: null,
                    liveUpdate: true
                },
                {
                    name: 'service-four',
                    module: 'module-three',
                    source: './services/modules/module-three/service-four.docker-compose.yaml',
                    dependsOn: [],
                    builder: '',
                    metadata: {
                        ingressRoute: 'Path("four")',
                        healthcheck: "ping me"
                    },
                    repository: null,
                    liveUpdate: true
                },
                {
                    name: 'service-five',
                    module: 'module-one',
                    source: './services/modules/module-one/service-five.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'java',
                    metadata: {
                        languageMajorVersion: '17',
                    },
                    repository: null,
                    liveUpdate: false
                },
                {
                    name: 'service-six',
                    module: 'module-one',
                    source: './services/modules/module-one/service-six.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'node',
                    metadata: {
                        languageMajorVersion: '18',
                        ingressRoute: 'Path("six")'
                    },
                    repository: null,
                    liveUpdate: false
                },
                {
                    name: 'service-seven',
                    module: 'module-two',
                    source: './services/modules/module-two/service-seven.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'repository',
                    metadata: {
                        repoContext: 'tilt/',
                        ingressRoute: 'Path("my-route")',
                    },
                    repository: null,
                    liveUpdate: true
                },
                {
                    name: 'service-eight',
                    module: 'module-two',
                    source: './services/modules/module-two/service-eight.docker-compose.yaml',
                    dependsOn: [],
                    builder: 'repository',
                    metadata: {
                        repoContext: 'tilt/',
                        ingressRoute: 'Path("my-route")',
                        healthcheck: 'curl http://localhost'
                    },
                    repository: null,
                    liveUpdate: true
                },
            ];

        it("correctly generates docker compose file, no exclusions", () => {
            dockerComposeFileGenerator.generateDockerComposeFile(
                services, []
            )

            const dockerComposeOutputFile = join(tempDir, "docker-compose.yaml")
            expect(existsSync(dockerComposeOutputFile)).toBe(true)

            const dockerComposeOutput = parse(readFileSync(dockerComposeOutputFile).toString('utf8'));

            dockerComposeOutput.include = dockerComposeOutput.include.map(
                (inclusion: string) => inclusion.replace(tempDir, ".")
            )

            expect(dockerComposeOutput).toMatchSnapshot();
        })

        it("correctly excludes services", () => {
            dockerComposeFileGenerator.generateDockerComposeFile(
                services, [
                "service-five",
                "service-six",
                "service-seven",
            ]
            )

            const dockerComposeOutputFile = join(tempDir, "docker-compose.yaml")
            expect(existsSync(dockerComposeOutputFile)).toBe(true)

            const dockerComposeOutput = parse(readFileSync(dockerComposeOutputFile).toString('utf8'));

            dockerComposeOutput.include = dockerComposeOutput.include.map(
                (inclusion: string) => inclusion.replace(tempDir, ".")
            )

            expect(dockerComposeOutput).toMatchSnapshot();
        })
    })

    describe("generateDevelopmentServiceDockerComposeFile", () => {
        let serviceDockerComposeFile: string;

        beforeEach(() => {
            serviceDockerComposeFile = join(tempDir, 'service.docker-compose.yaml');

            copyFileSync(
                join(process.cwd(), "test/data/docker-compose-file-generator/service.docker-compose.yaml"),
                serviceDockerComposeFile
            )

            rmSync(
                join(tempDir, "local/service-one"),
                {
                    force: true,
                    recursive: true
                }
            )
        })

        it("creates development docker compose correctly for latest node service", () => {
            const initialServiceDefinition = parse(readFileSync(serviceDockerComposeFile).toString("utf8"));

            initialServiceDefinition.services["service-one"].labels = [
                ...(initialServiceDefinition.services["service-one"].labels),
                "chs.local.builder=node"
            ]

            writeFileSync(
                serviceDockerComposeFile,
                stringify(initialServiceDefinition),
                {
                    flag: 'w'
                }
            )

            const service: Service = {
                name: 'service-one',
                module: 'module-one',
                source: serviceDockerComposeFile,
                repository: null,
                dependsOn: [],
                builder: 'node',
                metadata: {}
            }

            dockerComposeFileGenerator.generateDevelopmentServiceDockerComposeFile(service);

            const generatedDockerComposeFile = join(tempDir, "local/service-one/docker-compose.yaml");

            expect(existsSync(generatedDockerComposeFile)).toBe(true);

            const generatedDockerCompose = parse(readFileSync(generatedDockerComposeFile).toString("utf8"));

            generatedDockerCompose.services["service-one"].build.context = generatedDockerCompose.services["service-one"].build.context.replace(tempDir, "./");
            generatedDockerCompose.services["service-one"].build.dockerfile = generatedDockerCompose.services["service-one"].build.dockerfile.replace(tempDir, ".");
            generatedDockerCompose.services["service-one"].build.args["REPO_PATH"] = generatedDockerCompose.services["service-one"].build.args["REPO_PATH"].replace(tempDir, ".");

            expect(generatedDockerCompose).toMatchSnapshot();
        })

        it("creates development docker compose correctly node at specific major service", () => {
            const initialServiceDefinition = parse(readFileSync(serviceDockerComposeFile).toString("utf8"));

            initialServiceDefinition.services["service-one"].labels = [
                ...(initialServiceDefinition.services["service-one"].labels),
                "chs.local.builder=node",
                "chs.local.builder.languageVersion=18",
            ]

            writeFileSync(
                serviceDockerComposeFile,
                stringify(initialServiceDefinition),
                {
                    flag: 'w'
                }
            )

            const service: Service = {
                name: 'service-one',
                module: 'module-one',
                source: serviceDockerComposeFile,
                repository: null,
                dependsOn: [],
                builder: 'node',
                metadata: {
                    languageMajorVersion: "18"
                }
            }

            dockerComposeFileGenerator.generateDevelopmentServiceDockerComposeFile(service);

            const generatedDockerComposeFile = join(tempDir, "local/service-one/docker-compose.yaml");

            expect(existsSync(generatedDockerComposeFile)).toBe(true);

            const generatedDockerCompose = parse(readFileSync(generatedDockerComposeFile).toString("utf8"));

            generatedDockerCompose.services["service-one"].build.context = generatedDockerCompose.services["service-one"].build.context.replace(tempDir, "./");
            generatedDockerCompose.services["service-one"].build.dockerfile = generatedDockerCompose.services["service-one"].build.dockerfile.replace(tempDir, ".");
            generatedDockerCompose.services["service-one"].build.args["REPO_PATH"] = generatedDockerCompose.services["service-one"].build.args["REPO_PATH"].replace(tempDir, ".");

            expect(generatedDockerCompose).toMatchSnapshot();
        })

        it("creates development docker compose correctly for latest java service", () => {
            const initialServiceDefinition = parse(readFileSync(serviceDockerComposeFile).toString("utf8"));

            initialServiceDefinition.services["service-one"].labels = [
                ...(initialServiceDefinition.services["service-one"].labels),
                "chs.local.builder=java"
            ]

            writeFileSync(
                serviceDockerComposeFile,
                stringify(initialServiceDefinition),
                {
                    flag: 'w'
                }
            )

            const service: Service = {
                name: 'service-one',
                module: 'module-one',
                source: serviceDockerComposeFile,
                repository: null,
                dependsOn: [],
                builder: 'java',
                metadata: {}
            }

            dockerComposeFileGenerator.generateDevelopmentServiceDockerComposeFile(service);

            const generatedDockerComposeFile = join(tempDir, "local/service-one/docker-compose.yaml");

            expect(existsSync(generatedDockerComposeFile)).toBe(true);

            const generatedDockerCompose = parse(readFileSync(generatedDockerComposeFile).toString("utf8"));

            generatedDockerCompose.services["service-one"].build.context = generatedDockerCompose.services["service-one"].build.context.replace(tempDir, "./");
            generatedDockerCompose.services["service-one"].build.dockerfile = generatedDockerCompose.services["service-one"].build.dockerfile.replace(tempDir, ".");
            generatedDockerCompose.services["service-one"].build.args["REPO_PATH"] = generatedDockerCompose.services["service-one"].build.args["REPO_PATH"].replace(tempDir, ".");

            expect(generatedDockerCompose).toMatchSnapshot();
        })

        it("creates development docker compose correctly java at specific major service", () => {
            const initialServiceDefinition = parse(readFileSync(serviceDockerComposeFile).toString("utf8"));

            initialServiceDefinition.services["service-one"].labels = [
                ...(initialServiceDefinition.services["service-one"].labels),
                "chs.local.builder=java",
                "chs.local.builder.languageVersion=17",
            ]

            writeFileSync(
                serviceDockerComposeFile,
                stringify(initialServiceDefinition),
                {
                    flag: 'w'
                }
            )

            const service: Service = {
                name: 'service-one',
                module: 'module-one',
                source: serviceDockerComposeFile,
                repository: null,
                dependsOn: [],
                builder: 'java',
                metadata: {
                    languageMajorVersion: "17"
                }
            }

            dockerComposeFileGenerator.generateDevelopmentServiceDockerComposeFile(service);

            const generatedDockerComposeFile = join(tempDir, "local/service-one/docker-compose.yaml");

            expect(existsSync(generatedDockerComposeFile)).toBe(true);

            const generatedDockerCompose = parse(readFileSync(generatedDockerComposeFile).toString("utf8"));

            generatedDockerCompose.services["service-one"].build.context = generatedDockerCompose.services["service-one"].build.context.replace(tempDir, "./");
            generatedDockerCompose.services["service-one"].build.dockerfile = generatedDockerCompose.services["service-one"].build.dockerfile.replace(tempDir, ".");
            generatedDockerCompose.services["service-one"].build.args["REPO_PATH"] = generatedDockerCompose.services["service-one"].build.args["REPO_PATH"].replace(tempDir, ".");

            expect(generatedDockerCompose).toMatchSnapshot();
        })
    })
})
