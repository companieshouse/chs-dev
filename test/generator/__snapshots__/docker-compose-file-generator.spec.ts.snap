// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose correctly for latest java service 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "args": {
          "REPO_PATH": "./repositories/service-one",
        },
        "context": "./",
        "dockerfile": "./local/builders/java/Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=java",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose correctly for latest node service 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "args": {
          "REPO_PATH": "./repositories/service-one",
        },
        "context": "./",
        "dockerfile": "./local/builders/node/Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=node",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose correctly java at specific major service 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "args": {
          "LANGUAGE_MAJOR_VERSION": "17",
          "REPO_PATH": "./repositories/service-one",
        },
        "context": "./",
        "dockerfile": "./local/builders/java/Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=java",
        "chs.local.builder.languageVersion=17",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose correctly node at specific major service 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "args": {
          "LANGUAGE_MAJOR_VERSION": "18",
          "REPO_PATH": "./repositories/service-one",
        },
        "context": "./",
        "dockerfile": "./local/builders/node/Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=node",
        "chs.local.builder.languageVersion=18",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose correctly node with optional build args 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "args": {
          "ENTRYPOINT": "docker_start.sh",
          "LANGUAGE_MAJOR_VERSION": "18",
          "OUTDIR": "out/",
          "REPO_PATH": "./repositories/service-one",
        },
        "context": "./",
        "dockerfile": "./local/builders/node/Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "env_file": null,
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=node",
        "chs.local.builder.languageVersion=18",
        "chs.local.entrypoint=docker_start.sh",
        "chs.local.builder.outputDir=out/",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose for repository - with builder and dockerfile labels 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "context": ".//repositories/service-one",
        "dockerfile": "chs.Dockerfile",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=repository",
        "chs.local.dockerfile=chs.Dockerfile",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose for repository - with builder and repoContext labels 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "context": ".//repositories/service-one/chs-dev",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
        "chs.local.builder=repository",
        "chs.local.repoContext=chs-dev/",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDevelopmentServiceDockerComposeFile creates development docker compose for repository - with no builder label 1`] = `
{
  "services": {
    "service-one": {
      "build": {
        "context": ".//repositories/service-one",
      },
      "depends_on": [],
      "develop": {
        "watch": [
          {
            "action": "rebuild",
            "path": ".touch",
          },
        ],
      },
      "image": "my-great-image:latest",
      "labels": [
        "label-one=true",
      ],
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDockerComposeFile correctly excludes services 1`] = `
{
  "include": [
    "./services/modules/module-one/service-one.docker-compose.yaml",
    "./services/modules/module-one/service-two.docker-compose.yaml",
    "./local/service-three/docker-compose.yaml",
    "./local/service-four/docker-compose.yaml",
    "./services/modules/module-one/service-nine.docker-compose.yaml",
    "./local/service-eight/docker-compose.yaml",
  ],
  "services": {
    "ingress-proxy": {
      "depends_on": {
        "service-eight": {
          "condition": "service_healthy",
          "restart": true,
        },
        "service-four": {
          "condition": "service_healthy",
          "restart": true,
        },
        "service-nine": {
          "condition": "service_started",
          "restart": true,
        },
        "service-one": {
          "condition": "service_started",
          "restart": true,
        },
        "service-two": {
          "condition": "service_started",
          "restart": true,
        },
      },
      "image": "nginx:latest",
    },
  },
}
`;

exports[`DockerComposeFileGenerator generateDockerComposeFile correctly generates docker compose file, no exclusions 1`] = `
{
  "include": [
    "./services/modules/module-one/service-one.docker-compose.yaml",
    "./services/modules/module-one/service-two.docker-compose.yaml",
    "./local/service-three/docker-compose.yaml",
    "./local/service-four/docker-compose.yaml",
    "./services/modules/module-one/service-five.docker-compose.yaml",
    "./services/modules/module-one/service-six.docker-compose.yaml",
    "./services/modules/module-one/service-nine.docker-compose.yaml",
    "./local/service-seven/docker-compose.yaml",
    "./local/service-eight/docker-compose.yaml",
  ],
  "services": {
    "ingress-proxy": {
      "depends_on": {
        "service-eight": {
          "condition": "service_healthy",
          "restart": true,
        },
        "service-four": {
          "condition": "service_healthy",
          "restart": true,
        },
        "service-nine": {
          "condition": "service_started",
          "restart": true,
        },
        "service-one": {
          "condition": "service_started",
          "restart": true,
        },
        "service-seven": {
          "condition": "service_started",
          "restart": true,
        },
        "service-six": {
          "condition": "service_started",
          "restart": true,
        },
        "service-two": {
          "condition": "service_started",
          "restart": true,
        },
      },
      "image": "nginx:latest",
    },
  },
}
`;
